[Cocoa] Some AVIF images can't be rendered because their containers formats are HEIF

shallawa opened the pull request:
#### 08227904ddb6fdd6e7e85dd22928147d59b22879
<pre>
[Cocoa] Some AVIF images can&apos;t be rendered because their containers formats are HEIF
<a href="https://bugs.webkit.org/show_bug.cgi?id=244103">https://bugs.webkit.org/show_bug.cgi?id=244103</a>

Reviewed by Myles C. Maxfield.

If the system frameworks say the image type is &quot;public.heif&quot;, we can distingush
the case of a HEIF image from the case of an AVIF image inside a HEIF conatinver
by checking the &quot;ftyp&quot; box. An image is an AVIF if &quot;avif&quot; or &quot;avis&quot; is present
in the major_brand or in one of the compatible_brands of the &quot;ftyp&quot; box.

* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::decodeUTI const):
(WebCore::ImageDecoderCG::setData):
(WebCore::ImageDecoderCG::uti const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:

Canonical link: <a href="https://commits.webkit.org/254591@main">https://commits.webkit.org/254591@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70671e6621b826250b6164d0ca5806218554f7b6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89564 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34115 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98857 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155386 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93572 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32609 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28081 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/81920 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93267 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95211 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25900 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76424 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25847 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80778 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80651 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68837 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30366 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14728 "Found 1 new test failure: inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive.html (failure)") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30114 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15662 "Found 1 new test failure: inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive.html (failure)") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3227 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33563 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38616 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1351 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34753 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+    // An image is considered an AVIF if "avif" or "avis" is present in its
+    // major_brand or in one of the compatible_brands.
+
+    const unsigned* ftypBox = reinterpret_cast<const unsigned*>(data.data());
```
You never check `data.size()` so it might not have a 4 bytes in it (yet).

shallawa replied:
This code will be executed when` CGImageSourceGetType()` returns `public.heif`. And this can't happen if the `data.size()` is less than 4. So I added an ASSERT_NOT_REACHED().

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+    const unsigned* ftypBox = reinterpret_cast<const unsigned*>(data.data());
+
+    // The size is in little endian.
+    unsigned sizeInBytes = flipBytes(ftypBox[0]);
```
Let's call this `boxByteSize` (so as not confuse it with the image size).

Doesn't this also need to check if the code is complied for Big or LittleEndian? Better yet, use the functions in `ISOBox` since HEIF is based on the ISO format?

shallawa replied:
You are right. I made all the unsigned in Big Endian and I used FourCC to store the ftyp signatures and the avif brands.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+        return uti;
+    
+    constexpr auto bytesToUnsigned = [](const char bytes[4]) -> unsigned {
+        return bytes[0] | bytes[1] << 8 | bytes[2] << 16 | bytes[3] << 24;
```
Does this work if the code is compiled for a BigEndian architecture?

shallawa replied:
I used FourCC which is in Big Endian.

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+    if (ftypBox[1] != ftypSignature)
+        return uti;
+
+    unsigned sizeInUnsigned = sizeInBytes / sizeof(unsigned);
```
`boxUnsignedSize`?

shallawa replied:
Done

smfr commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +585,73 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+
+    unsigned sizeInUnsigned = sizeInBytes / sizeof(unsigned);
+
+    for (unsigned i = 2; i < sizeInUnsigned; ++i) {
```
Should we impose some reasonable limit so we don't walk through a potentially huge box size?

shallawa replied:
Done. I limit it to check only 100 compatible_brands.

smfr reviewed (commented):
We should add at least one testcase.

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+        return uti;
+
+    if (data.size() < sizeof(unsigned)) {
+        ASSERT_NOT_REACHED();
```
Why is this an assert? Servers could totally just reply with 3 bytes if they want

litherum replied:
Oh maybe the reason is because we are confident that CGImageSourceGetType() will never return heif or heic for any file that’s less than 4 bytes long?

Okay, that makes sense.

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+    
+    auto boxUnsigned = [&data](unsigned index) -> unsigned {
+        static constexpr bool littleEndian = false;
+        const unsigned* boxBytes = reinterpret_cast<const unsigned*>(data.data());
```
I think this is technically UB and probably should be done with bit manipulation instead.

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+        static constexpr bool littleEndian = false;
+        const unsigned* boxBytes = reinterpret_cast<const unsigned*>(data.data());
+        // Let's make all the numbers in big endian because FourCC are big endian.
+        return flipBytesIfLittleEndian(boxBytes[index], littleEndian);
```
Passing a constant false to a conditional function doesn’t make a lot of sense

shallawa replied:
This is for clarity only. Because the function takes a bool, I thought passing a constant `false` will not be clear.

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+    if (boxByteSize % 4 || boxByteSize > data.size())
+        return uti;
+
+    unsigned boxUnsignedSize = boxByteSize / sizeof(unsigned);
```
Isn’t it meaningless to do math on a big-endian number?

shallawa replied:
I am sorry for the confusion but I was confused myself by the little/big endian. Also the names of existing APIs do not help explain what I want to do. `flipBytesIfLittleEndian(num, false)` converts a big endian unsigned to match the Endianness of the current CPU architecture. So the math is correct in this case. And I fixed the comment that was saying "let's convert everything to big endian".

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -589,18 +586,96 @@ PlatformImagePtr ImageDecoderCG::createFrameImageAtIndex(size_t index, Subsampli
+
+    // Check compatible_brands.
+    while (boxBrandCount) {
+        if (auto avifUTI = checkAVIFBrand(boxUnsigned(boxUnsignedIndex++)))
```
Interesting. I suppose we’re confident that this won’t lead to false positives?

litherum commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
```diff
@@ -50,7 +50,7 @@ class ImageDecoderCG final : public ImageDecoder {
     size_t frameCount() const final;
     RepetitionCount repetitionCount() const final;
-    String uti() const final;
+    String uti() const final { return m_uti; }
```
Is it possible to add a test? If we know which encoder creates the problematic files, it seems like we could make a test image and commit it.

shallawa replied:
I added a ref test which is enabled only on macOS Ventura only for WK1. It is marked failed on WK2 because of the WKR sandboxing issue [webkit.org/b/240987](https://bugs.webkit.org/show_bug.cgi?id=240987).

shallawa committed:
[Cocoa] Some AVIF images can't be rendered because their containers formats are HEIF
https://bugs.webkit.org/show_bug.cgi?id=244103

Reviewed by Myles C. Maxfield.

If the system frameworks say the image type is "public.heif", we can distingush
the case of a HEIF image from the case of an AVIF image inside a HEIF conatinver
by checking the "ftyp" box. An image is an AVIF if "avif" or "avis" is present
in the major_brand or in one of the compatible_brands of the "ftyp" box.

* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::decodeUTI const):
(WebCore::ImageDecoderCG::setData):
(WebCore::ImageDecoderCG::uti const): Deleted.
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:

Canonical link: https://commits.webkit.org/254591@main
